### PR TITLE
support aggregation implementation

### PR DESCRIPTION
**Pull request by zybing**
Test plan - (Please fill in how you tested your changes)

support aggregation implementation
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by zybing** [comment]
@rongrong support java aggregate implementation, can you help review this code?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java [hunk 1]
```diff
@@ -302,4 +312,87 @@ public static Block extractMaskBlock(int maskChannel, Page page)
...
+                                          PagesIndex.Factory pagesIndexFactory,
+                                          boolean distinct,
+                                          JoinCompiler joinCompiler,
+                                          List<LambdaProvider> lambdaProviders,
+                                          boolean spillEnabled,
+                                          Session session,
+                                          StandaloneSpillerFactory standaloneSpillerFactory)
+    {
+        if (javaAggregationFunctionImplementation instanceof BuiltInAggregationFunctionImplementation) {
+            return ((BuiltInAggregationFunctionImplementation) javaAggregationFunctionImplementation).bind(
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java:331]: The only reason that you need an if instanceof here is the existence of `LazyAccumulatorFactoryBinder` which is not really needed. If you delete that, you can actually remove `AccumulatorFactoryBinder` from `JavaAggregationFunctionImplementation`, and create the binder class and call bind all here (the else block) for all `JavaAggregationFunctionImplementation`

File: presto-benchmark/src/main/java/com/facebook/presto/benchmark/CountAggregationBenchmark.java [hunk 2]
```diff
@@ -42,9 +43,10 @@ protected List<? extends OperatorFactory> createOperatorFactories()
     {
         OperatorFactory tableScanOperator = createTableScanOperator(0, new PlanNodeId("test"), "orders", "orderkey");
         FunctionAndTypeManager functionAndTypeManager = localQueryRunner.getMetadata().getFunctionAndTypeManager();
-        InternalAggregationFunction countFunction = functionAndTypeManager.getAggregateFunctionImplementation(
+        JavaAggregationFunctionImplementation countFunction = functionAndTypeManager.getJavaAggregateFunctionImplementation(
                 functionAndTypeManager.lookupFunction("count", fromTypes(BIGINT)));
-        AggregationOperatorFactory aggregationOperator = new AggregationOperatorFactory(1, new PlanNodeId("test"), Step.SINGLE, ImmutableList.of(countFunction.bind(ImmutableList.of(0), Optional.empty())), false);
+        AggregationOperatorFactory aggregationOperator = new AggregationOperatorFactory(1, new PlanNodeId("test"),
```
[the remark this answers is not in the document]
Comment (zybing) [reply presto-benchmark/src/main/java/com/facebook/presto/benchmark/CountAggregationBenchmark.java:48]: no logical change, just change the bind way. now using the `AggregationUtils.bind` method to bind function

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java [hunk 3]
```diff
@@ -423,6 +427,31 @@ public ScalarFunctionImplementation getScalarFunctionImplementation(FunctionHand
         return functionNamespaceManager.get().getScalarFunctionImplementation(functionHandle);
     }
 
+    public AggregationFunctionImplementation getAggregateFunctionImplementation(FunctionHandle functionHandle)
+    {
+        Optional<FunctionNamespaceManager<?>> functionNamespaceManager = getServingFunctionNamespaceManager(functionHandle.getCatalogSchemaName());
+        checkArgument(functionNamespaceManager.isPresent(), "Cannot find function namespace for '%s'", functionHandle.getCatalogSchemaName());
+        AggregationFunctionImplementation aggregationFunctionImplementation;
+        if (functionNamespaceManager.get().canResolveFunction()) {
```
[the remark this answers is not in the document]
Comment (zybing) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java:432]: yes, Thanks for your advice

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java [hunk 4]
```diff
@@ -423,6 +427,31 @@ public ScalarFunctionImplementation getScalarFunctionImplementation(FunctionHand
...
+        if (aggregationFunctionImplementation instanceof JavaAggregationFunctionImplementation) {
+            JavaAggregationFunctionImplementation functionImplementation = (JavaAggregationFunctionImplementation) aggregationFunctionImplementation;
+            if (functionImplementation.needConstructFactoryBinder()) {
+                GenericAccumulatorFactoryBinder factoryBinder = generateAccumulatorFactoryBinder(functionImplementation.getAggregationMetadata(),
+                        functionImplementation.getAccumulatorClass(), functionImplementation.getGroupedAccumulatorClass());
+                return new JavaAggregationFunctionImplementationAdapter(functionImplementation, Optional.of(factoryBinder));
+            }
+            return new JavaAggregationFunctionImplementationAdapter(functionImplementation, Optional.empty());
+        }
+        else {
```
[the remark this answers is not in the document]
Comment (zybing) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java]: outdated

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java [hunk 5]
```diff
@@ -302,4 +312,87 @@ public static Block extractMaskBlock(int maskChannel, Page page)
         }
         return maskBlock;
     }
+
+    public static AccumulatorFactory bind(JavaAggregationFunctionImplementation javaAggregationFunctionImplementation,
```
[the remark this answers is not in the document]
Comment (zybing) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java:316]: thanks, having change it

File: presto-benchmark/src/main/java/com/facebook/presto/benchmark/CountAggregationBenchmark.java [hunk 6]
```diff
@@ -46,7 +46,7 @@ protected List<? extends OperatorFactory> createOperatorFactories()
         JavaAggregationFunctionImplementation countFunction = functionAndTypeManager.getJavaAggregateFunctionImplementation(
                 functionAndTypeManager.lookupFunction("count", fromTypes(BIGINT)));
         AggregationOperatorFactory aggregationOperator = new AggregationOperatorFactory(1, new PlanNodeId("test"),
-                Step.SINGLE, ImmutableList.of(AggregationUtils.bind(countFunction, ImmutableList.of(0), Optional.empty())), false);
+                Step.SINGLE, ImmutableList.of(GenericAccumulatorFactory.generateAccumulatorFactory(countFunction, ImmutableList.of(0), Optional.empty())), false);
```
Comment (rongrong) [inline presto-benchmark/src/main/java/com/facebook/presto/benchmark/CountAggregationBenchmark.java:49]: nits: we prefer static import as long as the function name is not confusing, so here static import `GenericAccumulatorFactory.generateAccumulatorFactory` and this can just be `ImmutableList.of(generateAccumulatorFactory(countFunction, ImmutableList.of(0), Optional.empty()))`

Comment (rongrong) [reply presto-benchmark/src/main/java/com/facebook/presto/benchmark/CountAggregationBenchmark.java:49]: Please fix all other places to static import `GenericAccumulatorFactory.generateAccumulatorFactory` as well.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java [hunk 7]
```diff
@@ -145,6 +146,76 @@ public GenericAccumulatorFactory(
...
+                    distinct,
+                    spillEnabled,
+                    standaloneSpillerFactory);
+        }
+        catch (NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static AccumulatorFactory generateAccumulatorFactory(
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:199]: This one is probably only exist for testing purposes? Please annotate with `@VisibleForTesting` if that's the case.

File: presto-spi/src/main/java/com/facebook/presto/spi/function/JavaAggregationFunctionImplementation.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    default Class<? extends Accumulator> getAccumulatorClass()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    default Class<? extends GroupedAccumulator> getGroupedAccumulatorClass()
+    {
+        throw new UnsupportedOperationException();
+    }
```
Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/JavaAggregationFunctionImplementation.java:55]: I don't think you need to have default implementation for any of these functions.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 9]
```diff
@@ -2973,7 +2973,7 @@ private AccumulatorFactory buildAccumulatorFactory(
                     .map(LambdaDefinitionExpression.class::cast)
                     .collect(toImmutableList());
             for (int i = 0; i < lambdas.size(); i++) {
```
[the remark this answers is not in the document]
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:2979]: I prefer this so we can remove the `getLambdaInterfaces` from `JavaAggregationFunctionImplementation`

**Review by rongrong — COMMENTED** [review]
Please make the 3rd commit the first commit instead (refactor first before new implementation).

File: presto-benchmark/src/main/java/com/facebook/presto/benchmark/CountAggregationBenchmark.java (hunk 6 above)
Comment (zybing) [reply presto-benchmark/src/main/java/com/facebook/presto/benchmark/CountAggregationBenchmark.java:49]: 👌

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/BuiltInAggregationFunctionImplementation.java [hunk 10]
```diff
@@ -42,7 +38,71 @@
...
     private final boolean orderSensitive;
-    private final AccumulatorFactoryBinder factory;
+
+    private final AggregationMetadata aggregationMetadata;
+
+    private final Class<? extends Accumulator> accumulatorClass;
+
+    private final Class<? extends GroupedAccumulator> groupedAccumulatorClass;
+
+    public BuiltInAggregationFunctionImplementation(
```
[the remark this answers is not in the document]
Comment (zybing) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/BuiltInAggregationFunctionImplementation.java:48]: Try to simplify

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java (hunk 7 above)
Comment (zybing) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:199]: the methos is also used in `AggregateWindowFunction` class and `presto-benchmark` module

File: presto-spi/src/main/java/com/facebook/presto/spi/function/JavaAggregationFunctionImplementation.java (hunk 8 above)
Comment (zybing) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/JavaAggregationFunctionImplementation.java:55]: OK

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/CountColumn.java [hunk 11]
```diff
@@ -96,17 +98,16 @@ private static InternalAggregationFunction generateAggregation(Type type)
...
-
-        Class<? extends GroupedAccumulator> groupedAccumulatorClass = generateAccumulatorClass(
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
                 GroupedAccumulator.class,
                 metadata,
                 classLoader);
-        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), BIGINT, true,
-                false, metadata, accumulatorClass, groupedAccumulatorClass);
+        return new BuiltInAggregationFunctionImplementation(NAME, inputTypes, ImmutableList.of(intermediateType), BIGINT,
+                true, false, metadata, accumulatorClass, groupedAccumulatorClass);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/CountColumn.java:110]: unnecessary change.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java [hunk 12]
```diff
@@ -99,17 +101,16 @@ private static InternalAggregationFunction generateAggregation(Type keyType, Typ
...
-
-        Class<? extends GroupedAccumulator> groupedAccumulatorClass = generateAccumulatorClass(
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
                 GroupedAccumulator.class,
                 metadata,
                 classLoader);
-        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), outputType, true,
-                false, metadata, accumulatorClass, groupedAccumulatorClass);
+        return new BuiltInAggregationFunctionImplementation(NAME, inputTypes, ImmutableList.of(intermediateType), outputType,
+                true, false, metadata, accumulatorClass, groupedAccumulatorClass);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java:113]: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MergeStatisticalDigestFunction.java [hunk 13]
```diff
@@ -105,17 +107,16 @@ private InternalAggregationFunction generateAggregation(Type type)
...
-
-        Class<? extends GroupedAccumulator> groupedAccumulatorClass = generateAccumulatorClass(
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
                 GroupedAccumulator.class,
                 metadata,
                 classLoader);
-        return new InternalAggregationFunction(name, ImmutableList.of(type), ImmutableList.of(intermediateType), type, true,
-                true, metadata, accumulatorClass, groupedAccumulatorClass);
+        return new BuiltInAggregationFunctionImplementation(name, ImmutableList.of(type), ImmutableList.of(intermediateType), type,
+                true, true, metadata, accumulatorClass, groupedAccumulatorClass);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MergeStatisticalDigestFunction.java:119]: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/StatisticalDigestAggregationFunction.java [hunk 14]
```diff
@@ -150,17 +152,16 @@ private InternalAggregationFunction generateAggregation(String name, Type valueT
...
-
-        Class<? extends GroupedAccumulator> groupedAccumulatorClass = generateAccumulatorClass(
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
                 GroupedAccumulator.class,
                 metadata,
                 classLoader);
-        return new InternalAggregationFunction(name, inputTypes, ImmutableList.of(intermediateType), outputType, true,
-                true, metadata, accumulatorClass, groupedAccumulatorClass);
+        return new BuiltInAggregationFunctionImplementation(name, inputTypes, ImmutableList.of(intermediateType), outputType,
+                true, true, metadata, accumulatorClass, groupedAccumulatorClass);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/StatisticalDigestAggregationFunction.java:164]: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/ArrayAggregationFunction.java [hunk 15]
```diff
@@ -111,17 +111,17 @@ private static InternalAggregationFunction generateAggregation(Type type, boolea
...
-        Class<? extends GroupedAccumulator> groupedAccumulatorClass = generateAccumulatorClass(
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
                 GroupedAccumulator.class,
                 metadata,
                 classLoader);
-        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), outputType, true,
-                true, metadata, accumulatorClass, groupedAccumulatorClass);
+
+        return new BuiltInAggregationFunctionImplementation(NAME, inputTypes, ImmutableList.of(intermediateType), outputType,
+                true, true, metadata, accumulatorClass, groupedAccumulatorClass);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/ArrayAggregationFunction.java:124]: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java [hunk 16]
```diff
@@ -113,17 +113,16 @@ private static InternalAggregationFunction generateAggregation(Type type, ArrayT
...
-
-        Class<? extends GroupedAccumulator> groupedAccumulatorClass = generateAccumulatorClass(
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
                 GroupedAccumulator.class,
                 metadata,
                 classLoader);
-        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), outputType, true,
-                true, metadata, accumulatorClass, groupedAccumulatorClass);
+        return new BuiltInAggregationFunctionImplementation(NAME, inputTypes, ImmutableList.of(intermediateType),
+                outputType, true, true, metadata, accumulatorClass, groupedAccumulatorClass);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java:125]: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/Histogram.java [hunk 17]
```diff
@@ -107,17 +107,16 @@ private static InternalAggregationFunction generateAggregation(
...
-
-        Class<? extends GroupedAccumulator> groupedAccumulatorClass = generateAccumulatorClass(
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
                 GroupedAccumulator.class,
                 metadata,
                 classLoader);
-        return new InternalAggregationFunction(functionName, inputTypes, ImmutableList.of(intermediateType), outputType, true,
-                false, metadata, accumulatorClass, groupedAccumulatorClass);
+        return new BuiltInAggregationFunctionImplementation(functionName, inputTypes, ImmutableList.of(intermediateType),
+                outputType, true, false, metadata, accumulatorClass, groupedAccumulatorClass);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/Histogram.java:119]: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/minmaxby/AbstractMinMaxByNAggregationFunction.java [hunk 18]
```diff
@@ -171,16 +171,16 @@ protected InternalAggregationFunction generateAggregation(Type valueType, Type k
                         new MinMaxByNStateFactory())),
                 outputType);
 
-        Class<? extends Accumulator> accumulatorClass = generateAccumulatorClass(
+        Class<? extends Accumulator> accumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/minmaxby/AbstractMinMaxByNAggregationFunction.java:174]: unnecessary change / static import

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/multimapagg/MultimapAggregationFunction.java [hunk 19]
```diff
@@ -105,17 +105,16 @@ private InternalAggregationFunction generateAggregation(Type keyType, Type value
...
-
-        Class<? extends GroupedAccumulator> groupedAccumulatorClass = generateAccumulatorClass(
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
                 GroupedAccumulator.class,
                 metadata,
                 classLoader);
-        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), outputType, true,
-                true, metadata, accumulatorClass, groupedAccumulatorClass);
+        return new BuiltInAggregationFunctionImplementation(NAME, inputTypes, ImmutableList.of(intermediateType), outputType,
+                true, true, metadata, accumulatorClass, groupedAccumulatorClass);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/multimapagg/MultimapAggregationFunction.java:117]: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/operator/window/AggregateWindowFunction.java [hunk 20]
```diff
@@ -40,10 +42,12 @@
     private int currentStart;
     private int currentEnd;
 
-    private AggregateWindowFunction(InternalAggregationFunction function, List<Integer> argumentChannels)
+    private AggregateWindowFunction(AggregationFunctionImplementation function, List<Integer> argumentChannels)
     {
+        checkState(function instanceof BuiltInAggregationFunctionImplementation);
+        BuiltInAggregationFunctionImplementation buildinFunction = (BuiltInAggregationFunctionImplementation) function;
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/window/AggregateWindowFunction.java:48]: variable name: `builtinFunction`

File: presto-main/src/test/java/com/facebook/presto/operator/TestTableWriterOperator.java [hunk 21]
```diff
@@ -210,15 +210,15 @@ public void testStatisticsAggregation()
...
+        JavaAggregationFunctionImplementation longMaxFunction = functionAndTypeManager.getJavaAggregateFunctionImplementation(
                 functionAndTypeManager.lookupFunction("max", fromTypes(BIGINT)));
         TableWriterOperator operator = (TableWriterOperator) createTableWriterOperator(
                 pageSinkManager,
                 new AggregationOperatorFactory(
                         1,
                         new PlanNodeId("test"),
                         AggregationNode.Step.SINGLE,
-                        ImmutableList.of(generateAccumulatorFactory(longMaxFunction, ImmutableList.of(0), Optional.empty())),
+                        ImmutableList.of(GenericAccumulatorFactory.generateAccumulatorFactory(longMaxFunction, ImmutableList.of(0), Optional.empty())),
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/operator/TestTableWriterOperator.java:221]: static import

**Review by rongrong — APPROVED** [review]
Looks good. Just some final nits.

File: presto-main/src/test/java/com/facebook/presto/operator/TestTableWriterOperator.java (hunk 21 above)
Comment (zybing) [reply presto-main/src/test/java/com/facebook/presto/operator/TestTableWriterOperator.java]: has fixed it

File: presto-main/src/main/java/com/facebook/presto/operator/window/AggregateWindowFunction.java (hunk 20 above)
Comment (zybing) [reply presto-main/src/main/java/com/facebook/presto/operator/window/AggregateWindowFunction.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/multimapagg/MultimapAggregationFunction.java (hunk 19 above)
Comment (zybing) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/multimapagg/MultimapAggregationFunction.java:117]: It's too long for one line

File: presto-benchmark/src/main/java/com/facebook/presto/benchmark/DoubleSumAggregationBenchmark.java [hunk 22]
```diff
@@ -44,10 +44,10 @@ protected List<? extends OperatorFactory> createOperatorFactories()
     {
         OperatorFactory tableScanOperator = createTableScanOperator(0, new PlanNodeId("test"), "orders", "totalprice");
         FunctionAndTypeManager functionAndTypeManager = MetadataManager.createTestMetadataManager().getFunctionAndTypeManager();
-        InternalAggregationFunction doubleSum = functionAndTypeManager.getAggregateFunctionImplementation(
+        JavaAggregationFunctionImplementation doubleSum = functionAndTypeManager.getJavaAggregateFunctionImplementation(
                 functionAndTypeManager.lookupFunction("sum", fromTypes(DOUBLE)));
-        AggregationOperatorFactory aggregationOperator = new AggregationOperatorFactory(1, new PlanNodeId("test"),
-                Step.SINGLE, ImmutableList.of(generateAccumulatorFactory(doubleSum, ImmutableList.of(0), Optional.empty())), false);
+        AggregationOperatorFactory aggregationOperator = new AggregationOperatorFactory(1, new PlanNodeId("test"), Step.SINGLE,
```
[the remark this answers is not in the document]
Comment (zybing) [reply presto-benchmark/src/main/java/com/facebook/presto/benchmark/DoubleSumAggregationBenchmark.java]: done
